Tasks/Common/artifacts-common/connectionDataUtils.ts

```diff
@@ -0,0 +1,48 @@
+import * as tl from 'azure-pipelines-task-lib/task';
+import * as protocols from './protocols'
+import * as api from './webapi';
+import { ConnectOptions } from 'azure-devops-node-api/interfaces/common/VSSInterfaces';
+import { ConnectionData } from 'azure-devops-node-api/interfaces/LocationsInterfaces';
+
+/**
+ * Gets the raw connection data (direct representation of _apis/connectionData) for the service hosting a particular protocol
+ * @param protocolType The packaging protocol, e.g. 'NuGet'
+ */
+export async function getConnectionDataForProtocol(protocolType: protocols.ProtocolType) : Promise<ConnectionData> {
+    // Determine where the Packaging service lives
+    tl.debug('Finding the URI packaging service');
+    const accessToken = api.getSystemAccessToken();
+    const areaId = protocols.getAreaIdForProtocol(protocolType);
+    const serviceUri = await getServiceUriFromAreaId(areaId, accessToken); 
+    
+    // Get _apis/connectionData from the packaging service
+    const webApi = api.getWebApiWithProxy(serviceUri, accessToken);
+    const locationApi = await webApi.getLocationsApi();
+    tl.debug(`Acquiring connection data from ${serviceUri}`);
+    const connectionData = await locationApi.getConnectionData(ConnectOptions.IncludeServices);
+    tl.debug('Successfully acquired the connection data');
+
+    return connectionData;
+}
+
+/**
+ * Gets the URI of the service that hosts an area.
+ */
+async function getServiceUriFromAreaId(areaId: string, accessToken: string): Promise<string> {
+    const tfsCollectionUrl = tl.getVariable('System.TeamFoundationCollectionUri');
+    const serverType = tl.getVariable('System.ServerType');
+    if (!serverType || serverType.toLowerCase() !== 'hosted') {
+        return tfsCollectionUrl;
+    }
+
+    const webApi = api.getWebApiWithProxy(tfsCollectionUrl, accessToken);
+    const locationApi = await webApi.getLocationsApi();
+
+    tl.debug(`Getting URI for area ID ${areaId} from ${tfsCollectionUrl}`);
+    try {
+        const serviceUriFromArea = await locationApi.getResourceArea(areaId);
```

Comment:
When this errors out, it throws a really bad exception message. It would be good to add some additional logging here before throwing it.

Reply:
Yeah I just pulled this as-is from the existing utility, seemed odd to me to immediately re-throw like but I assumed there was a historical reason for it. I can definitely add additional logging.